### PR TITLE
Expanded the language-antlr-utils to allow the extraction of semantic…

### PR DESCRIPTION
**Pull request by TwoOfTwelve**
Adds the option to extract semantics information to the AbstractAntlrListener. For an example how to use, take a look at the tests in language-antlr-utils.

### CODE REVIEW

**Review by mbrdl — COMMENTED** [review]
I'm not sure how this framework pans out in practice, but I think I'll just make adjustements myself if needed.

**Comment by mbrdl** [comment]
I can't really make a call on the framework until I've tried using it. I think you can merge @tsaglam

File: language-antlr-utils/src/main/java/de/jplag/antlr/ContextTokenBuilder.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        } else {
+            return antlrContent.getStop().getStopIndex() - antlrContent.getStart().getStartIndex() + 1;
+        }
+    }
+
+    enum Type {
+        START,
+        STOP,
+        RANGE
+    }
```
Comment (tsaglam) [inline language-antlr-utils/src/main/java/de/jplag/antlr/ContextTokenBuilder.java:58]: Not the biggest fan of nested types, especially as this one is not private. Nested types tend to be improperly documented and harder to maintain.

File: language-antlr-utils/src/main/java/de/jplag/antlr/InternalListenerException.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package de.jplag.antlr;
+
+/**
+ * Exception type used internally within the antlr utils. Has to be a {@link RuntimeException}, because it is thrown
+ * within the antlr listener methods. Should not be thrown outside the antlr utils.
+ */
+public class InternalListenerException extends RuntimeException {
```
Comment (tsaglam) [inline language-antlr-utils/src/main/java/de/jplag/antlr/InternalListenerException.java:7]: The name of this exception does not really help with understand when this exception is thrown or what scenario it covers.

File: language-antlr-utils/src/main/java/de/jplag/antlr/SemanticsDataHolder.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package de.jplag.antlr;
+
+import de.jplag.semantics.VariableRegistry;
+
+/**
+ * Holds all required context objects for the generation of code semantics
+ * @param registry The variable registry
+ */
+public record SemanticsDataHolder(VariableRegistry registry) {
```
Comment (tsaglam) [inline language-antlr-utils/src/main/java/de/jplag/antlr/SemanticsDataHolder.java:9]: If this is a record encapsulating a single object, why do we need the record at all?

File: language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+import de.jplag.Token;
+import de.jplag.TokenType;
+import de.jplag.semantics.CodeSemantics;
+
+/**
+ * Handles the extraction of tokens. Contains information on the appropriate antlr types, the conditions under which the
+ * token should be extracted and semantics information.
+ * @param <T> The antlr type being mapped
+ */
+@SuppressWarnings("UnusedReturnValue")
```
Comment (tsaglam) [inline language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java:18]: Why do we suppress this? We should avoid doing so!

File: language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+public abstract class TokenBuilder<T> {
+    private static final Logger LOG = Logger.getLogger(TokenBuilder.class.getName());
```
Comment (tsaglam) [inline language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java:20]: Not in line with the usual naming in JPlag.

**Review by tsaglam — CHANGES_REQUESTED** [review]
Some issues we need to address.

File: language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java (hunk 4 above)
Comment (tsaglam) [reply language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java:18]: Yes, I think that is better.

File: language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+        int line = antlrToken.getLine();
+        int column = antlrToken.getCharPositionInLine() + 1;
+        int length = antlrToken.getText().length();
+
+        Token token;
+        if (semantics != null) {
+            if (semanticsSupplier == null) {
+                throw new InternalListenerException(MISSING_SEMANTICS);
+            }
```
[the remark this answers is not in the document]
Comment (tsaglam) [reply language-antlr-utils/src/main/java/de/jplag/antlr/TokenBuilder.java:120]: I would say, in that case you can just throw an IllegelStateException or a similar runtime exception and just not catch it. Parser exceptions are for the user, but the scenario you are describing here would affect the developer.

File: language-antlr-utils/src/main/java/de/jplag/antlr/SemanticsDataHolder.java (hunk 3 above)
Comment (tsaglam) [reply language-antlr-utils/src/main/java/de/jplag/antlr/SemanticsDataHolder.java:9]: Makes sense, but could be considered premature optimization of sorts. I am unsure if we should keep it in or not. @mbrdl do you have a preference?

Comment (mbrdl) [reply language-antlr-utils/src/main/java/de/jplag/antlr/SemanticsDataHolder.java:9]: The variable registry even started off as a semantics data holder, the name only came later.

Comment (tsaglam) [reply language-antlr-utils/src/main/java/de/jplag/antlr/SemanticsDataHolder.java]: @TwoOfTwelve then let us remove it for now! 👍

**Review by tsaglam — APPROVED** [review]

**Pull request merged**